src/VisualStudio/Core/Def/DocumentOutline/DocumentOutlineViewModel.cs

```diff
@@ -0,0 +1,149 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.ObjectModel;
+using System.ComponentModel;
+using System.Runtime.CompilerServices;
+using System.Threading;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Editor.Tagging;
+using Microsoft.CodeAnalysis.Navigation;
+using Microsoft.CodeAnalysis.Shared.TestHooks;
+using Microsoft.CodeAnalysis.Text;
+using Microsoft.VisualStudio.LanguageServer.Client;
+using Microsoft.VisualStudio.Text.Editor;
+using Roslyn.Utilities;
+
+namespace Microsoft.VisualStudio.LanguageServices.DocumentOutline
+{
+    /// <summary>
+    /// Responsible for updating data related to Document outline.
+    /// It is expected that all operations this type do not need to be on the UI thread.
+    /// The only exception is the constructor which should be initialized on the UI thread.
+    /// Any operations that do need to happen on a specific thread should be delegated out
+    /// to some other service such as <see cref="VisualStudioCodeWindowInfoService"/>.
+    /// </summary>
+    internal partial class DocumentOutlineViewModel : INotifyPropertyChanged, IDisposable
+    {
+        private readonly ILanguageServiceBroker2 _languageServiceBroker;
+        private readonly VisualStudioCodeWindowInfoService _visualStudioCodeWindowInfoService;
+        private readonly CompilationAvailableTaggerEventSource _textViewEventSource;
+        private readonly CancellationTokenSource _cancellationTokenSource;
+        private readonly IDocumentNavigationService _navigationService;
+        private readonly Workspace _workspace;
+        private CancellationToken CancellationToken => _cancellationTokenSource.Token;
+
+        public DocumentOutlineViewModel(
+            ILanguageServiceBroker2 languageServiceBroker,
+            IAsynchronousOperationListener asyncListener,
+            VisualStudioCodeWindowInfoService visualStudioCodeWindowInfoService,
+            CompilationAvailableTaggerEventSource textViewEventSource,
+            Workspace workspace,
+            IDocumentNavigationService documentNavigationService)
+        {
+            _languageServiceBroker = languageServiceBroker;
+            _visualStudioCodeWindowInfoService = visualStudioCodeWindowInfoService;
+            _textViewEventSource = textViewEventSource;
+            _workspace = workspace;
+            _navigationService = documentNavigationService;
+            _cancellationTokenSource = new CancellationTokenSource();
+
+            // initialize public properties
```

Comment:
not really a necessary comment. 

nit: just initialize these at the field location.  it's not specific to this constructor.